### PR TITLE
Range datatype

### PR DESCRIPTION
**Pull request by matthuhiggins**
Introduces support for date and integer ranges.

* Adds `:date_range` and `:integer_range` types. Each are represented as a Range object.
* An array tuple is stored in the database.
* Open ended integer ranges are represented in the database with `nil` and in the ruby Range object as Float::INFINITY. This hack will improve with ruby 2.6's native support for open-ended ranges.

### CODE REVIEW

File: lib/superstore/types/float_type.rb [hunk 1]
```diff
@@ -1,10 +1,6 @@
 module Superstore
   module Types
     class FloatType < BaseType
-      def encode(float)
-        float
-      end
-
```
Comment (dcromer) [inline lib/superstore/types/float_type.rb]: Seems unrelated to this PR.  You are probably good to pull this right now.

File: lib/superstore/types/range_type.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        {
+          lower: convert(:typecast, value[:lower] || value['lower']),
+          upper: convert(:typecast, value[:upper] || value['upper'])
+        }
+      end
+
+      private
+
+      def convert(method, value)
+        subtype.send(method, value) unless value.nil?
```
Comment (dcromer) [inline lib/superstore/types/range_type.rb]: `unless value.nil?` is unnecessary IMO.

File: lib/superstore/types/range_type.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+          convert(:encode, range_hash[:upper])
+        ]
+      end
+
+      def decode(range_tuple)
+        {
+          lower: convert(:decode, range_tuple[0]),
+          upper: convert(:decode, range_tuple[1])
+        }
+      end
```
Comment (dcromer) [inline lib/superstore/types/range_type.rb]: This doesn't make sense to me.  `decode` always accepts a string.  Typecast is used for non-string instantiation.

File: lib/superstore/types/range_type.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      #
+      def to_range_hash(value)
+        if value.is_a?(Range)
+          [value.begin, value.end]
+        elsif value.is_a?(Hash)
+          [value[:lower], value[:upper]]
+        elsif value.is_a?(Array)
+          [value[0], value[1]]
+        end
+      end
```
Comment (dcromer) [inline lib/superstore/types/range_type.rb]: I assume that superstore consumers are supposed to tie directly into this?

**Comment by dcromer** [comment]
💡 some tests for expected encode/decode/typecast behavior might make the expected behavior more obvious.

**Comment by matthuhiggins** [comment]
* encode/decode is the write/read to the postgres column.
* typecast is how the orm model converts input to the attribute accessor.

File: lib/superstore/types/range_type.rb (hunk 2 above)
Comment (matthuhiggins) [reply lib/superstore/types/range_type.rb]: That requires all types to check for nil in all methods, which is currently not required.

**Comment by matthuhiggins** [comment]
@dcromer I only shared this for you to base work off of, not as something to merge into master.

I am going to try making an actual ruby `Range` instance. I don't like the core ORM having this phony Hash definition.

**Comment by malept** [comment]
It's unfortunate timing that [Ruby 2.6](https://www.ruby-lang.org/en/news/2018/05/31/ruby-2-6-0-preview2-released/) will introduce "endless ranges":

> An endless range, `(1..)`, is introduced. It works as it has no end.

**Comment by matthuhiggins** [comment]
> It's unfortunate timing that Ruby 2.6 will introduce "endless ranges":

Wowzers thanks for the heads up. Makes me more convinced to use a ruby `Range`.

The storage design won't change, nor will the end-user json output, so we should be ok.

**Comment by matthuhiggins** [comment]
@dcromer @malept This is complete. A separate PR was done to clarify decode vs typecast, and that is merged into this.

**Comment by dcromer** [comment]
👁 👁

**Comment by dcromer** [comment]
I added some (literal) edge cases 😎 

My understanding is that [Postgres ranges](https://www.postgresql.org/docs/9.6/static/rangetypes.html) must be`[min_or_nil, max_or_nil]`, so:
    * Decode can always assume this format.
    * Encode must always encode to this format, including when provided `max..min` ruby ranges
    * Typecast can assume arrays are always `[min_or_nil, max_or_nil]`

**Comment by malept** [comment]
I don't think we should be constrained by the PostgreSQL range format, since the values are going into a JSONB object.

**Comment by dcromer** [comment]
You are right.  However I think we should still enforce [min_or_nil, max_or_nil].

File: lib/superstore/types/range_type.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+      def convert_min(method, value)
+        subtype.send(method, value)
+      end
+
+      def convert_max(method, value)
+        subtype.send(method, value)
+      end
+    end
```
Comment (matthuhiggins) [inline lib/superstore/types/range_type.rb]: No point in the ordered_range method. encode can use min/max at encode time.

There’s no use case for this so unsure what the expectation is.

File: test/unit/serialization_test.rb [hunk 6]
```diff
@@ -2,9 +2,12 @@
 
 class Superstore::SerializationTest < Superstore::TestCase
   test 'as_json' do
-    issue = Issue.new
-    expected = {"id" => issue.id}
+    issue = Issue.new(skill_level_required: 100..200)
+    expected = {
+      "id" => issue.id,
+      "skill_level_required" => [100, 200]
```
Comment (matthuhiggins) [inline test/unit/serialization_test.rb]: 👎 

Override `Range#as_json` in the app. See how we override Time#as_json for reference.
